### DISCUSSION TITLE
Support for floating point exceptional support

### DISCUSSION THREAD

**Discussion by xinyi-li7** [ROCm/ROCm#2332] [Q&A] [answered]
Hello,

I've been studying the [ISA guide](https://www.amd.com/system/files/TechDocs/instinct-mi100-cdna1-shader-instruction-set-architecture%C2%A0.pdf) and came across a section detailing that the AMD MI100 can intercept hardware exceptions. As stated:

> The AMD CDNA devices can detect floating point exceptions and can generate interrupts. In particular, they can detect IEEE floating-point exceptions in hardware; these can be recorded for post-execution analysis.

In light of this, I've made an attempt to facilitate exception handling in my own code by using the `-ffp-exception-behavior=strict` flag while compiling with hipcc. Unfortunately, despite setting this flag, I'm unable to access any floating-point exception information.

Here I attached a simple example where a division-by-zero occurred and, even with the aforementioned flag enabled, the program failed to report the error.

I'm trying to figure out if I'm using the correct flag or whether there might be something else I'm missing. Any insights on this would be deeply appreciated. Thank you!

The program is
```c++
#include "hip/hip_runtime.h"

#include <stdio.h>
#include <stdlib.h>


__global__ void dot_prod(float *x, float *y, int size)
{
  float d;
  for (int i=0; i < size; ++i)
  {
    float tmp;
    tmp = x[i]*y[i];
    tmp = (x[i]) / (x[i] - y[i]); // division by zero, produce inf
    d += tmp; // d=NaN
  }

  int tid = blockIdx.x * blockDim.x + threadIdx.x;
  if (tid == 0) {
    printf("dot: %f\n", d);
  }
}
int main(int argc, char **argv)
{
  int n = 3;
  int nbytes = n*sizeof(float);
  float *d_a = 0;
  hipMalloc(&d_a, nbytes);

  float *data = (float *)malloc(nbytes);
  for (int i=0; i < n; ++i)
  {
    data[i] = (float)(i+1);
  }

  hipMemcpy((void *)data, (void *)data, nbytes, hipMemcpyHostToDevice);

  printf("Calling kernel\n");
  dot_prod<<<1,1>>>(d_a, d_a, nbytes);
  hipDeviceSynchronize();
  printf("done\n");

  return 0;
}
```

**Comment by huanrwan-amd** [comment, marked as answer]
Hi @xinyi-li7, thank you for posting questions.
You can try adding these codes to main() to check exceptions explicitly:
```
  // Enable floating-point exceptions
  feenableexcept(FE_DIVBYZERO | FE_INVALID | FE_OVERFLOW);
```
Header file is
`#include <fenv.h> // For floating-point exceptions
`
The expected output would be:
```
Calling kernel
dot: -inf
done
```

**Comment by b-sumner** [comment]
Exceptions can be enabled on the GPU, but that causes the GPU to run slower and given there is just one exception bit of each type for the entire wave, it is hard to determine which thread(s) caused the exception.  If you're worried about division by zero, I'd suggest using a function that checks the denominator and possibly prints a message and then possibly abort()s.
